File: uoj_978_lemmings-battle.py
# By definition, we want a structure that maintains order (from highest to lowest), on every round.

from queue import PriorityQueue

n = int(input())

while n:
    b, sg, sb = [int(i) for i in input().split()]

    greenSoldiers = PriorityQueue()
    blueSoldiers = PriorityQueue()

    ## Reverse , to get descending
    for i in range(sg):
        greenSoldiers.put(- int(input()))

    for i in range(sb):
        blueSoldiers.put(- int(input()))
    
    while True:
        tmpblue = []
        tmpgreen = []
        for i in range(b):
            # Equivalent to blue - green, but reversed due to - inside of queue.
            if greenSoldiers.empty() or blueSoldiers.empty():
                break
            win =  greenSoldiers.get() - blueSoldiers.get()
            # if 0 both of them die
            if win > 0:
                tmpblue.append(-win)
            elif win < 0:
                tmpgreen.append(win)

        for i in tmpblue:
            blueSoldiers.put(i)
        for i in tmpgreen:
            greenSoldiers.put(i)

        if greenSoldiers.empty() or blueSoldiers.empty():
            break


    if greenSoldiers.empty() and blueSoldiers.empty():
        print("green and blue died")
    elif greenSoldiers.empty():
        print("blue wins")
        while not blueSoldiers.empty():
            print(- blueSoldiers.get())
    else:
        print("green wins")
        while not greenSoldiers.empty():
            print(- greenSoldiers.get())

    n -= 1
    if n:
        print()

File: uoj_978_lemmings-battle.cpp
#import <iostream>
#import <queue>
#import <vector>

using namespace std;

int main() {
  int n; // number of test cases
  cin >> n;

  int b,sg,sb; // battle positions, greensoldiers, bluesoldiers;
  int tmp, win; // placeholders, avoid reinitialization

  while (n--) {
    cin >> b >> sg >> sb;

    priority_queue<int> greenSoldiers;
    priority_queue<int> blueSoldiers;

    for (int i = 0; i < sg; i++) {
      cin >> tmp;
      greenSoldiers.push(tmp);
    }

    for (int i = 0; i < sb; i++) {
      cin >> tmp;
      blueSoldiers.push(tmp);
    }

    // Finished with input handling

    while (true)
      {
	vector<int> tmpblue;
	vector<int> tmpgreen;

	for (int i = 0; i < b; i++)
	  {
	    if ((blueSoldiers.empty())||(greenSoldiers.empty()))
	      break;

	    win = blueSoldiers.top() - greenSoldiers.top();
	    blueSoldiers.pop();
	    greenSoldiers.pop();

	    if (win > 0)
	      tmpblue.push_back(win);
	    else if (win < 0)
	      tmpgreen.push_back(-win);
	  }

	for (int i = 0; i < tmpblue.size(); i++)
	  blueSoldiers.push(tmpblue[i]);
	for (int i = 0; i < tmpgreen.size(); i++)
	  greenSoldiers.push(tmpgreen[i]);

	if ((blueSoldiers.empty())||(greenSoldiers.empty()))
	    break;
      }
    if ((blueSoldiers.empty())&&(greenSoldiers.empty()))
      cout << "green and blue died" << endl;
    else if (greenSoldiers.empty())
      {
	cout << "blue wins" << endl;
	while (not (blueSoldiers.empty()))
	  {
	    cout << blueSoldiers.top() << endl;
	    blueSoldiers.pop();
	  }
      }
    else // if (blueSoldiers.empty())
      {
	cout << "green wins" << endl;
	while (not (greenSoldiers.empty()))
	  {
	    cout << greenSoldiers.top() << endl;
	    greenSoldiers.pop();
	  }
      }
    if (n)
      cout << endl;


  }
}
